### PR TITLE
Document GitLab with private container registries

### LINKED ISSUE #387: Document usage of Docker credential helpers on GitLab
* [x] Amazon Web Services
* [x] Google Cloud Platform
* [ ] Microsoft Azure

# Examples

_Loosely based on the pertaining [official documentation](https://docs.gitlab.com/ee/ci/docker/using_docker_images.html#use-credential-helpers)_

## Amazon Web Services

* Moved from https://github.com/iterative/terraform-provider-iterative/issues/627
* May be used with an instance profile as per https://github.com/iterative/terraform-provider-iterative/issues/602#issuecomment-1145964625

### GitLab CI/CD Environment Variables
| Name | Value |
| -- | -- |
| `DOCKER_AUTH_CONFIG` | `{"credHelpers": {"ACCOUNT.dkr.ecr.REGION.amazonaws.com": "ecr-login"}}` |
| `AWS_ACCESS_KEY_ID` | [AWS access key identifier](https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html) |
| `AWS_SECRET_ACCESS_KEY` | [AWS secret access key](https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html) |
| `AWS_SESSION_TOKEN` | [AWS session token ***(optional)***](https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html) |
| `REPO_TOKEN` | [GitLab Personal Access Token](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html) | 

### `.gitlab-ci.yml`
```yaml
runner:
  when: always
  image: iterativeai/cml
  script:
  - cml runner --labels=cml --cloud=aws

job:
  tags: [cml]
  needs: [runner]
  image: ACCOUNT.dkr.ecr.REGION.amazonaws.com/REPOSITORY:TAG
  script:
  - echo succeeded
```

> **Note**
> Replace the `ACCOUNT`, `REGION`, `REPOSITORY` and `TAG` placeholders with appropriate values.

## Google Cloud Platform

* Moved from https://github.com/iterative/cml/issues/1136#issuecomment-1240108523

### GitLab CI/CD Environment Variables
| Name | Value |
| -- | -- |
| `DOCKER_AUTH_CONFIG` | `{"credHelpers": {"LOCATION-docker.pkg.dev": "gcr"}}` |
| `GOOGLE_APPLICATION_CREDENTIALS_DATA` | [Contents of a Google Cloud service account JSON key file](https://cloud.google.com/iam/docs/creating-managing-service-account-keys#creating) |
| `REPO_TOKEN` | [GitLab Personal Access Token](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html) | 

### `.gitlab-ci.yml`
```yaml
runner:
  when: always
  image: iterativeai/cml
  script:
  - cml runner --labels=cml --cloud=gcp --cloud-permission-set="$(printenv GOOGLE_APPLICATION_CREDENTIALS_DATA | jq .client_email),scopes=storage-ro,datastore"

job:
  tags: [cml]
  needs: [runner]
  image: LOCATION-docker.pkg.dev/PROJECT/REPOSITORY/IMAGE:TAG
  script:
  - echo succeeded
```

> **Note**
> Replace the [`LOCATION`](https://cloud.google.com/artifact-registry/docs/repositories/repo-locations), `PROJECT`, `REPOSITORY`, `IMAGE` and `TAG` placeholders with appropriate values.

> **Warning**
> This example uses the new [Artifact Registry](https://cloud.google.com/artifact-registry) (i.e. any of the `LOCATION-docker.pkg.dev` domains) instead of the old Container Registry (i.e. the `gcr.io` domain) but instructions are similar for both.

## Microsoft Azure

⚠️ **To be done**

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
Closes #387, opens #413

### CODE REVIEW

File: content/docs/self-hosted-runners.md [hunk 1]
```diff
@@ -433,7 +433,95 @@ provisioned through environment variables instead of files.
 </tab>
 </toggle>
 
-### On-premise (Local) Runners
+## GitLab CI/CD and container images from private registries
+
+_See also the
+[official documentation](https://docs.gitlab.com/ee/ci/docker/using_docker_images.html#use-credential-helpers)
```
Comment (jorgeorpinel) [inline content/docs/self-hosted-runners.md:439]:
```suggestion
[GitLab docs](https://docs.gitlab.com/ee/ci/docker/using_docker_images.html#use-credential-helpers)
```

File: content/docs/self-hosted-runners.md [hunk 2]
```diff
@@ -433,7 +433,95 @@ provisioned through environment variables instead of files.
...
+for more information._
+
+### Amazon Web Services
+
+#### GitLab CI/CD Environment Variables
```
Comment (jorgeorpinel) [inline content/docs/self-hosted-runners.md:444]: Too much structure? Just plain sentences would work too. Or use tabs like the existing sections?

**Comment by jorgeorpinel** [comment]
LGTM but can't confirm the correctness.

File: content/docs/self-hosted-runners.md [hunk 3]
```diff
@@ -448,7 +448,101 @@ The same credentials can also be used for
 
 </admon>
 
-### On-premise (Local) Runners
+## GitLab CI/CD and container images from private registries
```
Comment (casperdcl) [inline content/docs/self-hosted-runners.md:451]: this is GL specific? what about GH?

File: content/docs/self-hosted-runners.md [hunk 4]
```diff
@@ -448,7 +448,101 @@ The same credentials can also be used for
...
+[Artifact Registry](https://cloud.google.com/artifact-registry) (i.e. any of the
+`LOCATION-docker.pkg.dev` domains) instead of the old Container Registry (i.e.
+the `gcr.io` domain) but instructions are similar for both.
+
+</admon>
+
+</tab>
+</toggle>
+
+## On-premise (Local) Runners
```
[the remark this answers is not in the document]
Comment (0x2b3bfa0) [reply content/docs/self-hosted-runners.md:545]: Arguably, the previous level did not make any sense. How is `On-premise (Local) Runners` supposed to be a child of the `Environment Variables` section? 🤔

File: content/docs/self-hosted-runners.md (hunk 3 above)
Comment (0x2b3bfa0) [reply content/docs/self-hosted-runners.md:451]: This is GitLab-specific. On GitHub it's not possible.[^1]

The closest approximation I can think of is this (not working) example:

```yaml
on: push
jobs:
  authentication:
    runs-on: ubuntu-latest
    outputs:
      username: ${{ steps.authenticate.outputs.username }}
      password: ${{ steps.authenticate.outputs.password }}
    steps:
      - id: authenticate
        run: |
          echo "username=AWS" >> $GITHUB_OUTPUT
          echo "password=$(aws ecr get-login-password --region us-east-1)" >> $GITHUB_OUTPUT
  example:
    needs: authentication
    runs-on: ubuntu-latest
    container:
      image: node
      credentials:
        username: ${{ needs.authentication.outputs.username }}
        password: ${{ needs.authentication.outputs.password }}
    steps:
      - run: true
```

[^1]: Read as «not easy enough to be documented»

Comment (0x2b3bfa0) [reply content/docs/self-hosted-runners.md:451]: The closest approximation I can think of is this (not working) example:

```yaml
on: push
jobs:
  authentication:
    runs-on: ubuntu-latest
    outputs:
      username: ${{ steps.authenticate.outputs.username }}
      password: ${{ steps.authenticate.outputs.password }}
    steps:
      - id: authenticate
        run: |
          echo "username=AWS" >> $GITHUB_OUTPUT
          echo "password=$(aws ecr get-login-password --region us-east-1)" >> $GITHUB_OUTPUT
  example:
    needs: authentication
    runs-on: ubuntu-latest
    container:
      image: node
      credentials:
        username: ${{ needs.authentication.outputs.username }}
        password: ${{ needs.authentication.outputs.password }}
    steps:
      - run: true
```

Comment (0x2b3bfa0) [reply content/docs/self-hosted-runners.md:451]: ...or https://agileek.github.io/software/aws/2020/10/28/using-an-ecr-image-in-github-actions

Comment (casperdcl) [reply content/docs/self-hosted-runners.md:451]: kewl. I think this deserves a new page as it's not really about self-hosted runners, right?

Comment (0x2b3bfa0) [reply content/docs/self-hosted-runners.md:451]: It _is_ about self-hosted runners, on GitLab, with private container registries. Definition of niceh.

**Comment by 0x2b3bfa0** [comment]
🔔 @casperdcl: meow! 🐱

File: content/docs/self-hosted-runners.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (casperdcl) [reply content/docs/self-hosted-runners.md:545]: moved to #446 for a quickfix :)

**Comment by shcheklein** [comment]
@0x2b3bfa0 can we merge this?

**Comment by 0x2b3bfa0** [comment]
Yes! @shcheklein, can you please approve it?

**Review by shcheklein — APPROVED** [review]

**Pull request merged**